os/drivers/usbhost/uvc/usbhost_uvc_ctrl.c

```diff
@@ -0,0 +1,1360 @@
+		 * UVC controls that support it.
+		 */
+		if (mapping->data_type == UVC_CTRL_DATA_TYPE_BITMASK && (ctrl->info.flags & F_GET_RES)) {
+			ret = uvc_ctrl_populate_values(chain, ctrl);
+			if (ret < 0) {
+				udbg("Failed to populate control values!!\n");
+				return ret;
+			}
+
+			step = uvc_get_value(mapping, UVC_GET_RES, ctrl->step);
+			if (!(step & value)) {
+				uvdbg("Failed to get setps from control\n");
+				return -EINVAL;
+			}
+		}
+
+		break;
+
+	default:
+		value = xctrl->value;
+		break;
+	}
+
+	if (!CTRL_IS_LOADED(ctrl->status) && (ctrl->info.size * 8) != mapping->size) {
+		if ((ctrl->info.flags & F_GET_CUR) == 0) {
+			memset(ctrl->cur, 0, ctrl->info.size);
+		} else {
+			ret = uvc_query_ctrl(chain->dev, UVC_GET_CUR, ctrl->entity->id, chain->dev->intfnum, ctrl->info.selector, ctrl->cur, ctrl->info.size);
+			if (ret < 0) {
+				udbg("Failed to query the control(%d) to device!!\n", ctrl->info.selector);
+				return ret;
+			}
+			ret = OK;
+		}
+
+		CTRL_SET_LOADED(ctrl->status);
+	}
+
+	/* Backup the current value in case we need to rollback later. */
+	if (!CTRL_IS_DIRTY(ctrl->status)) {
+		memcpy(ctrl->bak, ctrl->cur, ctrl->info.size);
+	}
+
+	uvc_set_value(mapping, value, ctrl->cur);
+
+	CTRL_SET_DIRTY(ctrl->status);
+	CTRL_SET_MODIFIED(ctrl->status);
+	return 0;
+}
+
+/****************************************************************************
+ * Name: uvc_ctrl_restore_values
+ *
+ * Description:
+ *   This function restore the value on resume from a low power cycle.
+ *
+ *   This function is called from the resume method of driver.
+ *
+ * Input Parameters:
+ *   dev - reference to the driver object.
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+int uvc_ctrl_restore_values(uvc_state_t *dev)
+{
+	struct uvc_control *ctrl;
+	struct uvc_entity *entity = queue_entry(dev->entities.head, struct uvc_entity, list);
+	unsigned int i;
+	int ret = OK;
+
+	/* Walk the entities list and restore controls when possible. */
+	for (; &entity->list != (dq_entry_t *)&dev->entities; entity = queue_entry(entity->list.flink, struct uvc_entity, list)) {
+
+		for (i = 0; i < entity->ncontrols; ++i) {
+			ctrl = &entity->controls[i];
+
+			if (!CTRL_IS_SUPPORTED(ctrl->status) || !CTRL_IS_MODIFIED(ctrl->status) || (ctrl->info.flags & F_RESTORE) == 0) {
+				continue;
+			}
+
+			uvdbg("restoring control %pUl/%u/%u\n", ctrl->info.entity, ctrl->info.index, ctrl->info.selector);
+			CTRL_SET_DIRTY(ctrl->status);
+
+			if (ctrl->info.flags & F_AUTO || !(ctrl->info.flags & F_GET_CUR)) {
+				CTRL_CLEAR_LOADED(ctrl->status);
+			}
+
+			ret = uvc_query_ctrl(dev, UVC_SET_CUR, ctrl->entity->id, dev->intfnum, ctrl->info.selector, ctrl->cur, ctrl->info.size);
+
+			if (ret < 0) {
+				memcpy(ctrl->cur, ctrl->bak, ctrl->info.size);
+			} else {
+				ret = OK;
+			}
+
+			CTRL_CLEAR_DIRTY(ctrl->status);
```

Comment:
regardless ret, it should set as a dirty here?

Reply:
Yes, this is flag which is used to mark the current values are dirty before restoring and after restoring the values the status is marked as not dirty. This is used in power resumes or USB suspend/resume case.